src/Fabulous.XamarinForms/VirtualizedCollection.fs

```diff
@@ -0,0 +1,82 @@
+        
+        let onBindingContextChanged () =
+            match target.BindingContext with
+            | null -> ()
+            | value ->
+                let currWidget =
+                    match value with
+                    | :? Widget as widget -> widget
+                    | :? GroupItem as groupItem when isHeader = true -> groupItem.Header
+                    | :? GroupItem as groupItem when isHeader = false -> groupItem.Footer
+                    | v -> failwith $"Unexpected {v.GetType().FullName} in BindingContext"
+                    
+                let node = ViewNode.get target
+                Reconciler.update canReuseView prevWidgetOpt currWidget node
+                prevWidgetOpt <- ValueSome currWidget
+
+        onBindingContextChanged
+
+/// Create a DataTemplate for a specific root type (TextCell, ViewCell, etc.)
+/// that listen for BindingContext change to apply the Widget content to the cell
+type WidgetDataTemplate(``type``, isHeader, parent: IViewNode) =
+    inherit DataTemplate(fun () ->
+        let bindableObject = Activator.CreateInstance ``type`` :?> BindableObject
+        
+        let viewNode = ViewNode(ValueSome parent, parent.TreeContext, WeakReference(bindableObject))
+        bindableObject.SetValue(ViewNode.ViewNodeProperty, viewNode)
+        
+        let onBindingContextChanged = BindableHelpers.createOnBindingContextChanged parent.TreeContext.CanReuseView isHeader bindableObject
+        bindableObject.BindingContextChanged.Add (fun _ -> onBindingContextChanged ())
+        
+        bindableObject :> obj
+    )
+
+/// Redirect to the right type of DataTemplate based on the target type of the current widget cell
+type WidgetDataTemplateSelector internal (node: IViewNode, isHeader: bool, getWidget: obj -> Widget) =
+    inherit DataTemplateSelector()
+    
+    /// Reuse data template for already known widget target type
+    let cache = Dictionary<Type, WidgetDataTemplate>()
+    
+    override _.OnSelectTemplate(item, _) =
+        let widget = getWidget item
+        let widgetDefinition = WidgetDefinitionStore.get widget.Key
+        let targetType = widgetDefinition.GetTargetType(widget)
+        match cache.TryGetValue(targetType) with
+        | true, dataTemplate -> dataTemplate
+        | false, _ ->
+            let dataTemplate = WidgetDataTemplate(targetType, isHeader, node)
+            cache.Add(targetType, dataTemplate)
+            dataTemplate
+
+type SimpleWidgetDataTemplateSelector(node: IViewNode) =
+    inherit WidgetDataTemplateSelector(node, false, fun item -> item :?> Widget)
+        
+type GroupedWidgetDataTemplateSelector(node: IViewNode, isHeader: bool) =
+    inherit WidgetDataTemplateSelector(node, isHeader, fun item ->
+        let groupItem = item :?> GroupItem
+        if isHeader then groupItem.Header else groupItem.Footer
+    )
+        
+type WidgetItems =
+    { OriginalItems: IEnumerable
```

Comment:
can we somehow provide a type there? My understanding that it is a very old (.net 1.1?) collection that stores each item boxed. Probably not an issue. I'm wondering what can be stored here in practice?

Reply:
Actually `IEnumerable` is an interface implemented by all collections (`IEnumerable<T>` inherit from it), so it will not store boxed items, but will box on enumeration.

And since `WidgetItems` is the value that is stored inside `Attribute`, I don't think we can type it more. The list is the user provided one, not something we have control of.

Reply:
I'm wondering if I could do something like:
```fs
type WidgetItems<'T> =
    { OriginalItems: IEnumerable<'T>
       Template: 'T -> Widget }
       
module ItemsView =
    let WidgetItemsSource<'T> =
        Attributes.defineBindableWithComparer<WidgetItems<'T>, WidgetItems<'T>, IEnumerable<Widget>>
            Xamarin.Forms.ItemsView.ItemsSourceProperty
            id
            (fun modelValue -> seq { for x in modelValue.OriginalItems do modelValue.Template x })
            (fun (a, b) -> ScalarAttributeComparers.equalityCompare(a.OriginalItems, b.OriginalItems))
```

This will instantiate an attribute definition by type of `'T`, but maybe better than boxing all items on each enumeration?

Reply:
I like the typed one, but I will defer the final judgement to you :)

Reply:
I like the typed one too. And I think it's better to pay for a small one time startup cost along with a small increase in app size, rather than slowing down scrolling in lists